stage1/init/common/mount.go

```diff
@@ -68,131 +69,260 @@ func (m *Mount) Source(podRoot string) string {
-	// So, we generate a pile of Mounts and their corresponding Volume
+	mp, ok := s.mounts[vol]
+	return mp, ok
+}
 
-	// Map of hostpath -> Mount
-	mnts := make(map[string]schema.Mount)
+// volumeOf returns a volume details of the POD with an indicator of
+// volume existence.
+func (s *rtmState) volumeOf(vol types.ACName) (types.Volume, bool) {
+	if s.volumes == nil {
+		s.volumes = make(volumesMap, len(s.rtm.Volumes))
+		for _, v := range s.rtm.Volumes {
+			s.volumes[v.Name] = v
+		}
+	}
 
-	// Check runtimeApp's Mounts
-	for _, m := range ra.Mounts {
-		mnts[m.Path] = m
+	v, ok := s.volumes[vol]
+	return v, ok
+}
+
+// RuntimeMounts represents a mount points generator of the runtime
+// application.
+type RuntimeMounts struct {
+	// Application is the runtime application to generate mount points.
+	Application *schema.RuntimeApp
+
+	// Volumes specifies a list of volumes of the application.
+	Volumes []types.Volume
+
+	// DockerImplicit indicates if the resulting mount points should be
+	// marked as implicit empty volumes from a Docker image.
+	DockerImplicit bool
+}
+
+// NewRuntimeMounts creates a new instance of the RuntimeMounts with a
+// attributes retrieved from the given POD configuration.
+func NewRuntimeMounts(ra *schema.RuntimeApp, pod *stage1types.Pod) *RuntimeMounts {
+	return &RuntimeMounts{
+		Application: ra,
+		Volumes:     pod.Manifest.Volumes,
+		DockerImplicit: ConvertedFromDocker(
+			pod.Images[ra.Name.String()],
+		),
+	}
+}
+
+// emptyVolume returns a new instance of the empty volume with a unique
+// name and default configuration.
+func (rtm *RuntimeMounts) emptyVolume(mpoint *types.MountPoint) types.Volume {
+	defaultMode := "0755"
+	uniqName := rtm.Application.Name + "-" + mpoint.Name
+
+	return types.Volume{
+		Name: uniqName,
+		Kind: "empty",
+		Mode: &defaultMode,
+		UID:  new(int),
+		GID:  new(int),
+	}
+}
 
+// traverseMounts builds a mapping of the mounts of the runtime application
+// to the volumes declared in a POD manifest. For each successfully created
+// mapping function executes a specified call-back function for further
+// processing.
+//
+// An error returns, when runtime application declares a mount for non-existing
+// volume (i.e. POD manifest is missing the respective volume declaration).
+func (rtm *RuntimeMounts) traverseMounts(s *rtmState, f func(m Mount) error) error {
+	// Check runtimeApp's Mounts
+	for _, m := range rtm.Application.Mounts {
 		vol := m.AppVolume // Mounts can supply a volume
 		if vol == nil {
-			vv, ok := vols[m.Volume]
+			vv, ok := s.volumeOf(m.Volume)
 			if !ok {
-				return nil, fmt.Errorf("could not find volume %s", m.Volume)
+				return fmt.Errorf("could not find volume %s", m.Volume)
 			}
 			vol = &vv
 		}
 
 		// Find a corresponding MountPoint, which is optional
-		ro := false
-		for _, mp := range ra.App.MountPoints {
-			if mp.Name == m.Volume {
-				ro = mp.ReadOnly
-				break
-			}
-		}
+		mpoint, _ := s.mountPointOf(m.Volume)
```

Comment:
This function returns instance of ```types.MountPoint```, which is a structure and default value for it is an *empty* structure. Therefore, when mount point was not found, it returns ```types.MountPoint{}```, where ```ReadOnly``` attribute is set to ```false``` - a default value for boolean type.

The answer is ```mpoint``` is never going to be equal to ```nil```.